clients/programhash.go

```diff
@@ -0,0 +1,130 @@
+
+import (
+	"bytes"
+	"encoding/json"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/NethermindEth/juno/core/crypto"
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+type ContractCode struct {
+	Abi     interface{} `json:"abi"`
+	Program *Program    `json:"program"`
+}
+
+func ProgramHash(contractDefinition *ClassDefinition) (*felt.Felt, error) {
+	program := contractDefinition.Program
+
+	// make debug info None
+	program.DebugInfo = nil
+
+	// Cairo 0.8 added "accessible_scopes" and "flow_tracking_data" attribute fields, which were
+	// not present in older contracts. They present as null/empty for older contracts deployed
+	// prior to adding this feature and should not be included in the hash calculation in these cases.
+	//
+	// We, therefore, check and remove them from the definition before calculating the hash.
+	if program.Attributes != nil {
+		attributes := program.Attributes.([]interface{})
+		if len(attributes) == 0 {
+			program.Attributes = nil
+		} else {
+			for key, attribute := range attributes {
+				attributeInterface := attribute.(map[string]interface{})
+				if attributeInterface["accessible_scopes"] == nil || len(attributeInterface["accessible_scopes"].([]interface{})) == 0 {
+					delete(attributeInterface, "accessible_scopes")
+				}
+
+				if attributeInterface["flow_tracking_data"] == nil || len(attributeInterface["flow_tracking_data"].(map[string]interface{})) == 0 {
+					delete(attributeInterface, "flow_tracking_data")
+				}
+
+				attributes[key] = attributeInterface
+			}
+
+			program.Attributes = attributes
+		}
+	}
+
+	contractCode := new(ContractCode)
+	contractCode.Abi = contractDefinition.Abi
+	contractCode.Program = &program
+
+	programBytes, err := contractCode.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	programKeccak, err := crypto.StarkNetKeccak(programBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return programKeccak, nil
+}
+
+// Marshal is a custom json marshaler for ContractCode
+func (c ContractCode) Marshal() ([]byte, error) {
+	contractCodeBytes, err := json.Marshal(c)
+	if err != nil {
+		return nil, err
+	}
+
+	// reformat hints
+	hint := c.Program.Hints
+	newHint := reformatHint(hint)
+
+	contractCodeString := string(contractCodeBytes)
+	hintsIndex := strings.Index(contractCodeString, `"hints":`)            // find location of "hints" in contractCodeString
+	identifierIndex := strings.Index(contractCodeString, `"identifiers":`) // find location of "identifiers" in contractCodeString
+	// replace json "hints" with reformatted "hints"
+	contractCodeString = contractCodeString[:(hintsIndex+8)] + newHint + contractCodeString[identifierIndex-1:]
+	// reformat contractCodeString with
+	contractCodeString = strings.ReplaceAll(contractCodeString, `,`, `, `)
+	contractCodeString = strings.ReplaceAll(contractCodeString, `,  `, `, `)
+	contractCodeString = strings.ReplaceAll(contractCodeString, `:`, `: `)
+	contractCodeString = strings.ReplaceAll(contractCodeString, `:  `, `: `)
```

Comment:
These could easily cause problems as those characters could be used anywhere, not just as delimiters. Lets merge and see if this ends up being a problem.

also we should use escaped whitespace characters (`\r`, `\n`, `\t` etc), it is not easily recognizable what the whitespaces are in string constants.